### PR TITLE
Add type annotations to GroupBy objects

### PR DESCRIPTION
**Pull request by xinrong-databricks**
`#type: ignore`'s in b20a3b1 are mostly because mypy cannot identify `_attributes` of `self`.

We are considering adding abstract methods to `GroupBy` to help mypy identify. However, this should be a separate task.

### CODE REVIEW

File: databricks/koalas/groupby.py [hunk 1]
```diff
@@ -2061,7 +2075,7 @@ def pandas_transform(pdf):
 
         return DataFrame(internal)
 
-    def nunique(self, dropna=True):
+    def nunique(self, dropna=True) -> DataFrame:
```
Comment (ueshin) [inline databricks/koalas/groupby.py:2078]: `Union[DataFrame, Series]`?

File: databricks/koalas/series.py [hunk 2]
```diff
@@ -5888,7 +5888,7 @@ def unpack_scalar(sdf):
     return l2[0]
 
 
-def first_series(df):
+def first_series(df) -> "Series":
```
Comment (ueshin) [inline databricks/koalas/series.py:5891]: Actually it's `Union["Series", pd.Series]`.

Comment (ueshin) [reply databricks/koalas/series.py:5891]: I guess we can just remove pandas DataFrame support here.

File: databricks/koalas/groupby.py (hunk 1 above)
Comment (xinrong-databricks) [reply databricks/koalas/groupby.py]: Modified!

File: databricks/koalas/groupby.py [hunk 3]
```diff
@@ -1808,18 +1818,22 @@ def head(self, n=5):
...
         else:
             agg_columns = [
                 kdf._kser_for(label)
                 for label in kdf._internal.column_labels
-                if label not in self._column_labels_to_exlcude
+                if label not in self._column_labels_to_exlcude  # type: ignore
             ]
 
-        kdf, groupkey_labels, _ = self._prepare_group_map_apply(kdf, self._groupkeys, agg_columns)
+        kdf, groupkey_labels, _ = self._prepare_group_map_apply(
```
[the remark this answers is not in the document]
Comment (xinrong-databricks) [reply databricks/koalas/groupby.py]: Good catch! Modified.

File: databricks/koalas/groupby.py [hunk 4]
```diff
@@ -303,7 +307,7 @@ def _spark_groupby(kdf, func, groupkeys=()):
             data_spark_columns=[scol_for(sdf, col) for col in data_columns],
         )
 
-    def count(self):
+    def count(self) -> DataFrame:
```
[the remark this answers is not in the document]
Comment (xinrong-databricks) [reply databricks/koalas/groupby.py]: Makes sense, modified!

File: databricks/koalas/series.py (hunk 2 above)
Comment (xinrong-databricks) [reply databricks/koalas/series.py:5891]: I see. Let me change it to `Union["Series", pd.Series]` first.

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.

**Pull request merged**